File: backend/app/schemas/record_links.py
"""Contracts for the Phase G.1 record-to-parcel association workflow."""

from __future__ import annotations

import uuid
from datetime import datetime
from typing import Any, Literal

from pydantic import BaseModel, Field, field_validator

from app.schemas.common import PageMetadata


LinkStatus = Literal["SUGGESTED", "REVIEW_REQUIRED", "CONFIRMED", "REJECTED"]
LinkMethod = Literal["EXACT_SURVEY_IDENTIFIER", "ATTRIBUTE_MATCH", "SPATIAL_CONTEXT", "MANUAL"]


class LinkSuggestionRequest(BaseModel):
    document_validation_result_id: uuid.UUID


class ManualLinkRequest(LinkSuggestionRequest):
    parcel_id: uuid.UUID
    rationale: str = Field(min_length=1, max_length=2000)

    @field_validator("rationale")
    @classmethod
    def require_rationale(cls, value: str) -> str:
        if not value.strip():
            raise ValueError("Manual link rationale must not be blank.")
        return value


class LinkResolutionRequest(BaseModel):
    reason: str | None = Field(default=None, max_length=2000)


class RecordParcelLinkResponse(BaseModel):
    id: uuid.UUID
    project_id: uuid.UUID
    document_id: uuid.UUID
    document_validation_result_id: uuid.UUID
    parcel_id: uuid.UUID
    parcel_display_identifier: str | None
    link_status: LinkStatus
    link_method: LinkMethod
    confidence: float | None
    rationale: dict[str, Any]
    provenance: dict[str, Any]
    review_required: bool
    review_task_id: uuid.UUID | None
    created_by_user_id: uuid.UUID
    reviewed_by_user_id: uuid.UUID | None
    reviewed_at: datetime | None
    review_reason: str | None
    created_at: datetime
    updated_at: datetime


class RecordParcelLinkListResponse(BaseModel):
    items: list[RecordParcelLinkResponse]
    page: PageMetadata


class RecordParcelLinkSuggestionResponse(BaseModel):
    items: list[RecordParcelLinkResponse]

File: frontend/src/api/recordLinks.ts
import { sessionFetch } from "./session";
import { ApiError } from "./gis";

export type RecordParcelLinkStatus = "SUGGESTED" | "REVIEW_REQUIRED" | "CONFIRMED" | "REJECTED";
export type RecordParcelLinkMethod = "EXACT_SURVEY_IDENTIFIER" | "ATTRIBUTE_MATCH" | "SPATIAL_CONTEXT" | "MANUAL";

export interface RecordParcelLink {
  id: string;
  project_id: string;
  document_id: string;
  document_validation_result_id: string;
  parcel_id: string;
  parcel_display_identifier: string | null;
  link_status: RecordParcelLinkStatus;
  link_method: RecordParcelLinkMethod;
  confidence: number | null;
  rationale: Record<string, unknown>;
  provenance: Record<string, unknown>;
  review_required: boolean;
  review_task_id: string | null;
  review_reason: string | null;
  created_at: string;
  updated_at: string;
}

interface Page<T> { items: T[]; page: { limit: number; offset: number; total: number }; }

async function request<T>(path: string, init?: RequestInit): Promise<T> {
  const token = sessionStorage.getItem("access_token");
  const response = await fetch(`${baseUrl}/api/v1${path}`, {
    ...init,
    headers: {
      ...(token ? { Authorization: `Bearer ${token}` } : {}),
      ...init?.headers,
    },
  });
  if (!response.ok) {
    const body = await response.json().catch(() => null) as { error?: { code?: string; message?: string } } | null;
    throw new ApiError(response.status, body?.error?.code ?? "RECORD_LINK_REQUEST_FAILED", body?.error?.message ?? "Unable to load record-to-parcel links.");
  }
  return response.json() as Promise<T>;
}

export function loadDocumentRecordLinks(projectId: string, documentId: string): Promise<Page<RecordParcelLink>> {
  return request<Page<RecordParcelLink>>(`/projects/${projectId}/documents/${documentId}/record-parcel-links?limit=100`);
}

export function suggestDocumentRecordLinks(projectId: string, documentId: string, validationResultId: string): Promise<{ items: RecordParcelLink[] }> {
  return request(`/projects/${projectId}/documents/${documentId}/record-parcel-links/suggestions`, {
    method: "POST",
    headers: { "Content-Type": "application/json" },
    body: JSON.stringify({ document_validation_result_id: validationResultId }),
  });
}

export function resolveRecordParcelLink(projectId: string, linkId: string, action: "confirm" | "reject", reason?: string): Promise<RecordParcelLink> {
  return request(`/projects/${projectId}/record-parcel-links/${linkId}/${action}`, {
    method: "POST",
    headers: { "Content-Type": "application/json" },
    body: JSON.stringify({ reason: reason?.trim() || null }),
  });
}

export function loadParcelRecordLinks(projectId: string, parcelId: string): Promise<Page<RecordParcelLink>> {
  return request<Page<RecordParcelLink>>(`/projects/${projectId}/parcels/${parcelId}/record-parcel-links?limit=100`);
}
